Feature/unit tests

dxL1nus opened the pull request:
closes #70
closes #40 
closes #73 
closes codecheckers/check-pub#35
closes #96

**!IMPORTANT:** This PR can only be safely merged after PR #63 

This is the PR for the first unit tests for the ojs-codecheck plugin. The tests are mainly for the RetrieveReserveIdentifiers scripts and can act as a guideline/ template on how to write other unit tests for the remaining class files.

Run the tests with the following command from the PKP OJS root directory:
```bash
lib/pkp/lib/vendor/phpunit/phpunit/phpunit -c lib/pkp/tests/phpunit.xml plugins/generic/codecheck/tests/
```

Results:
<img width="1360" height="728" alt="grafik" src="https://github.com/user-attachments/assets/de989727-2da9-4e1e-809c-e46ad1498987" />

find the test coverage in the following file:
```bash
/var/www/html/ojs-tests/lib/pkp/tests/results/index.html
```

<img width="1910" height="562" alt="grafik" src="https://github.com/user-attachments/assets/9aa15f21-f66c-40c7-8fc0-f205e7f65578" />

dxL1nus committed:
Initial structure for unittests plus first tests for getRawIdentifier #70

dxL1nus committed:
Found command to run unittests #70

dxL1nus committed:
Changed behavior of getRawIdentifier based on unittests #70

dxL1nus committed:
Added new unit-test for getRawIdentifier() #70

dxL1nus committed:
Update README.md to include Guide on how to run Unit Tests #70

dxL1nus committed:
CertificateIdentifierList is now fully covered #70

dxL1nus committed:
Simplified setup of CertificateIdentifierList tests #70

dxL1nus committed:
Added Unit Tests for CertificateIdentifier #70

dxL1nus committed:
Added further Unit tests for Codecheck Venues #70

dxL1nus committed:
Added some more UnitTests for the GitHub Api Parser #70

dxL1nus committed:
Added last UnitTests to reach 100% coverage #70

dxL1nus committed:
Merged main and adapted to latest version of RetrieveReserveIdentifier #70

dxL1nus committed:
Added unit test for Unique Array from factory method #70

dxL1nus committed:
Fixed Merge conflict in Readme Test Documentation #70

dxL1nus committed:
Added test coverage file path to README #70

dxL1nus committed:
Merged new commits to Unit Tests #70

nuest commented:
@Subhanaliweb will include these changes in his upcoming PR with a full test suite.

subhanLabs committed:
feat: add comprehensive unit and component tests

- Add PHP unit tests for all classes (212 tests)
  * Constants, Submission, Settings, Workflow, Frontend, Migration
  * CodecheckPlugin with comprehensive hook testing
  * 185 tests pass locally, 27 skipped (require full OJS environment)

- Add Cypress component tests (51 tests - all passing)
  * CodecheckManifestFiles (6 tests)
  * CodecheckMetadataForm (20 tests)
  * CodecheckRepositoryList (6 tests)
  * CodecheckReviewDisplay (7 tests)
  * CodecheckCertificateIdentifier (6 tests)
  * CodecheckDataAndSoftwareAvailability (6 tests)

- Add test infrastructure
  * PHPUnit: bootstrap.php, PKPTestCase.php, phpunit.xml
  * Cypress: cypress.config.js, pkp-mock.js, component support files
  * GitHub Actions workflow for automated component testing

- Update README with testing documentation
  * Local testing instructions
  * Docker/CI testing instructions
  * Component testing commands

Total: 263 tests (212 PHP + 51 component)

subhanLabs commented:
As discussed, I pulled this branch and applied my testing on top of the commits he made.

nuest commented:
> As discussed, I pulled this branch and applied my testing on top of the commits he made.

@Subhanaliweb Can you briefly confirm: 1. we can close this PR unmerged. 2. You will open a new PR with your combined test suites.

subhanLabs commented:
I have appleid my commits of all other testing on this branch so we can merge this one, my other branches will no longer be needed.

nuest commented:
Thanks for the clarification @Subhanaliweb.

@dxL1nus @Subhanaliweb Please explain why this PR is still a "Draft". Who is best suited to resolve the merge conflicts?

dxL1nus commented:
> Thanks for the clarification @Subhanaliweb.
> 
> @dxL1nus @Subhanaliweb Please explain why this PR is still a "Draft". Who is best suited to resolve the merge conflicts?

this needs to stay a draft for now as well. I have to dicuss with Subhan what to do with the methods from the `CodecheckMetadataHandler`, as I moved them a couple months ago to the `CodecheckApiController`, but in his workflow, they are still inside the `CodecheckMetadataHandler`

dxL1nus committed:
Updated most of the outdated tests #70

dxL1nus committed:
Merge branch 'main' into feature/unit-tests

dxL1nus committed:
Fixed remaining bug in CodecheckRegister unit tests #70

dxL1nus committed:
Fixed Cypress Certificate Identifier tests #35

dxL1nus committed:
Fixed Cypress Certificate Identifier tests https://github.com/codecheckers/check-pub/issues/35

dxL1nus committed:
Fixed DataAndSoftwareAvailability Textarea Cypress test https://github.com/codecheckers/check-pub/issues/35

dxL1nus committed:
Added Component tests CI status badge to README.md https://github.com/codecheckers/check-pub/issues/35

subhanLabs committed:
fix: update tests and source to match refactored class structure

- Make getAuthors and buildYaml public in CodecheckMetadataHandler
- Refactor CodecheckApiHandler to delegate to handler methods
- Update CodecheckSubmissionUnitTest to match new schema
- Update CodecheckSubmissionDAOUnitTest to match new schema
- Fix CodecheckMetadataHandlerUnitTest: getAuthors/buildYaml now public
- Fix PHPUnit command path in README

subhanLabs commented:
I've moved the methods back from `CodecheckApiHandler` to `CodecheckMetadataHandler` 
and updated `CodecheckApiHandler` to delegate to the handler methods. 
This aligns with the original design discussed. PR is ready for review and merge.

nuest committed:
Fix comment typo in CodecheckVenueNames.php

nuest commented on tests/SubmissionUnitTests/CodecheckSubmissionUnitTest.php:
This class mostly just tests get and set functions and constructors, these are blurring the actual tests that test functions.

subhanLabs replied:
You're right, I had too many getter/setter tests here. I've removed all the constructor and basic getter/setter tests, keeping only the ones that test actual logic like JSON decoding, conditional checks, and URL building. Went from 40+ tests down to 9.

nuest commented on tests/CodecheckPluginUnitTest.php:
- Here again, too basic things about the class are tested (return types, functions existance, ...), hiding the actual valuable tests. This is just one example.

subhanLabs replied:
Agreed, removed all the method existence and return type checks. Kept only the tests that verify actual plugin behavior like schema modifications, form field additions, and data saving logic. Reduced from 48 tests to 7.

nuest commented on tests/ConstantsUnitTest.php:
Do not test constants, test functionalities that can break or behave differently based on the input.

nuest replied:
Bumping the same comment - no need to test constants directly. They are tested by tests assuming these are defaults when using the actual plugin features.

dxL1nus replied:
I have removed most of the simple constant tests. There are only one or two tests remaining, which I thought seem quite useful. The where done by @Subhanaliweb, maybe he can help here as well.

subhanLabs replied:
Removed the entire file. Agreed that constants are already tested through the actual feature tests.

nuest committed:
Reformat

nuest commented on tests/CodecheckRegisterUnitTests/CertificateIdentifierListUnitTest.php:
```diff
@@ -0,0 +1,227 @@
+    {
+        $title = '     ';
+        $rawIdentifier = CertificateIdentifierList::getRawIdentifier($title);
+        // The rawIdentifier for an empty string should be null
```
Repeated same comment below often, and it provides no additional information to the code statements. Remove these comments, if you feel like you have to comment > improve the wording of the function name.

dxL1nus replied:
I removed all comments in the `CodecheckRegisterUnitTests`, as the test names should already be descriptive enough

nuest commented on tests/CodecheckRegisterUnitTests/CertificateIdentifierListUnitTest.php:
```diff
@@ -0,0 +1,227 @@
+        $this->assertSame($actualIdentifierListString, $expectedIdentifierListString);
+    }
+
+    public function testFromApiWithMockApiParserThrowsApiException()
```
This is basically a test of the mock function - what is the valuable tested code here? That the "fromApi" function internally calls "fetchIssues" ?

dxL1nus replied:
Not quite. When `CertificateIdentifierList::fromApi($apiParserMock);` is called to create a new `CertificateIdentifierList` it uses the `CodecheckGithubRegisterApiClient`. I want to test/ ensure, that the possible exceptions which can be thrown in `CodecheckGithubRegisterApiClient->fetchIssues()` are registered an caught.

So yes, in a way it is saying `'fetchIssues'` should throw an Exception, so now expect that Exception. But in a way, that other method is tested, which calls `'fetchIssues'` internally and now should also throw that Exception. Let me know if we want to test these kind of things or not.

nuest replied:
Since the mock immediately throws that exception, I do not see this as a valuable test. That would be different when the mock-setup leads to a situation, where some logic/code of your code runs into an Exception.

dxL1nus replied:
ok then I'll remove all the tests featuring such a exception mocking workflow

dxL1nus replied:
should be done now, please look over the CodecheckRegister Tests and confirm, that no useless tests exist now in your eyes.

dxL1nus committed:
#70, #79 : I deleted all unnecessary tests and created a sh script to execute the tests

dxL1nus committed:
#70, #79 : Added the results folder to the .gitignore

dxL1nus committed:
#70, #79 : Better README for how to run tests

dxL1nus committed:
#70, #79 : Fixed small bug in runTests.sh

dxL1nus committed:
Small fix in README

dxL1nus commented:
@nuest 
I removed every useless test and also created a file: `runTests.sh`, to be able to run tests locally without executing the very long command. It also has a flag with which the testers can specify, if they want PHPUnit to also create a test coverage report (in HTML format) or not.

See [here](https://github.com/dxL1nus/ojs-codecheck/tree/feature/unit-tests?tab=readme-ov-file#php-unit-tests) for the full instructions on how to run tests.

nuest commented on .phpunit.cache/code-coverage/00b7a9638976680d9cdead19a11604a0:
Please ignore the content of the folder `.phpunit.cache`. Revert the commit where the files were added, if it is not too recent.

dxL1nus replied:
Yes so this was my mistake, but until now I didn't notice it. I wrongly turned on cacheResult="true" in the `phpunit.xml`, when running the tests once, because that's what the configuration of OJS does as well. But we don't need it cached anyways, so I just removed the folder and the flag from the `phpunit.xml` all together. I'll still add it to the `.gitignore` though, in case someone creates the commit again

nuest commented on tests/FrontEndUnitTests/ArticleDetailsUnitTest.php:
```diff
@@ -0,0 +1,258 @@
+        $this->assertFalse($result);
+    }
+
+    public function testAddCodecheckInfoMethodExists()
```
I scrolled over quite a few of these tests, so let me make a comment at least.

When the function does not exist anymore (e.g., it was removed by a developer, or renamed) then the feature that uses this function will break - and it is that feature that we want to test.

Similar case: using reflection to check the names of the parameters of a function. If these change unexpectedly, the feature that uses this function will break - and that is what we should test.

I can live with them being there. I cannot see the value of them, I fear they are costly to maintain.

subhanLabs replied:
Removed all the method existence and parameter checks. Only the constructor test remains.

nuest commented on tests/FrontEndUnitTests/ArticleDetailsUnitTest.php:
```diff
@@ -0,0 +1,258 @@
+        $this->assertSame('string', $returnType->getName());
+    }
+
+    public function testGenerateSidebarDisplayAcceptsCorrectParameters()
```
Just one example of many: if you want to check that a function does what it should, then call the function.

I fear these reflection-based tests are going to be annoying when code is refactored during development.

subhanLabs replied:
Addressed above ↑

nuest commented on tests/MigrationUnitTests/CodecheckSchemaMigrationUnitTest.php:
We are testing the database structure sufficiently by testing the classes that read from and write to this database.

Is there one aspect in the tests here that will not also be caught by saving a check to DB and reading it back out?

subhanLabs replied:
Good point, removed the entire migration test file. The schema is already validated by the DAO tests when they save and retrieve data.

nuest commented on tests/SubmissionUnitTests/CodecheckSubmissionDAOUnitTest.php:
```diff
@@ -0,0 +1,154 @@
+            'repository' => 'https://github.com/new/repo'
+        ]);
+
+        $this->assertTrue(true);
```
This is not a valid test statement, this is always true.

subhanLabs replied:
Fixed, replaced assertTrue(true) with actual assertions that verify the insert/update operations return the correct values and the data is properly stored.

nuest commented on tests/SubmissionUnitTests/CodecheckSubmissionDAOUnitTest.php:
```diff
@@ -0,0 +1,154 @@
+            'repository' => 'https://github.com/updated/repo'
+        ]);
+
+        $this->assertTrue(true);
```
Same here - always true - please explain.

nuest commented on tests/SubmissionUnitTests/SchemaUnitTest.php:
Also seems like a very "meta" test class to me: what is tested here is not covered by the schema being used in the core plugin functions.

nuest commented on tests/WorkflowUnitTests/CodecheckMetadataHandlerUnitTest.php:
Mostly testing based on Reflection - what is tested here that is note implicitly tested through the core plugin functions?

subhanLabs replied:
Removed the entire file. The schema validation is already covered by the DAO tests that actually use these properties.

nuest commented on tests/runTests.sh:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env sh
+
+# Standard-Command
```
Only add comments that add information , `DEFAULT_CMD` includes all information from the comment.

dxL1nus replied:
Done in commit [730f7d8](https://github.com/codecheckers/ojs-codecheck/pull/79/commits/730f7d86862f15a5a581a00651ae481b63bc8208)

nuest commented on tests/runTests.sh:
```diff
@@ -0,0 +1,26 @@
+        exec $DEFAULT_CMD
+        ;;
+    *)
+        # unbekannter Parameter -> Standard
```
this line comment can stay - just translate to English, please

dxL1nus replied:
Done in commit [730f7d8](https://github.com/codecheckers/ojs-codecheck/pull/79/commits/730f7d86862f15a5a581a00651ae481b63bc8208)

nuest reviewed (changes requested):
Primary change: remove the erroneously committed cache files, ideally also from the history.

Other changes: new extensive test patterns that I do not understand the benefit of; these can stay if it allows us to move on, but I would appreciate the help to understand the reasoning.

dxL1nus committed:
#96 users now get notified when label retrieval did not work

dxL1nus committed:
Removed the wrongly commited phpunit cache files #79

dxL1nus committed:
#79 added .phpunit.cache to the .gitignore

dxL1nus committed:
Removed unnecessary constants unit tests #79

dxL1nus committed:
Removed unnecessary comments in runTests.sh #79

dxL1nus committed:
Correct URL for Codecheck API again

nuest commented:
@Subhanaliweb Please take a look at the unresolved comments above - they are related to test classes you created. Thanks!

dxL1nus committed:
Removed Bugs from main

dxL1nus committed:
Merge branch 'main' into feature/unit-tests

dxL1nus committed:
Merged main to feature/unit-tests #79

subhanLabs committed:
Remove meta-tests: constants, schema, and migration tests

- Removed ConstantsUnitTest.php (testing constants is not behavioral)
- Removed SchemaUnitTest.php (schema tested by DAO tests)
- Removed CodecheckSchemaMigrationUnitTest.php (migration meta-test)

subhanLabs committed:
Refactor unit tests to focus on functional testing

- Removed meta-tests (method existence, constants, return types)
- Kept only tests that verify actual behavior and logic
- Fixed assertion issues in DAO tests

subhanLabs committed:
Update PHPUnit configuration for coverage reports

subhanLabs committed:
Fix CodecheckManifestFiles component tests

- Use CSS class selectors instead of placeholder text
- Makes tests independent of translation mock

subhanLabs committed:
Downgrade Vite to 5.4.0 for Cypress compatibility

- Cypress doesn't support Vite 8 yet
- Vite 5.4.0 is the latest compatible version

subhanLabs committed:
Add comprehensive GitHub Actions CI/CD workflow

- Runs PHPUnit tests with full OJS environment (stable-3_5_0)
- Runs Cypress component tests in parallel
- Generates code coverage reports with Xdebug
- Uploads artifacts (coverage, logs, screenshots)
- Uses GitHub secrets for API credentials
- Replaces old component-tests.yml with unified workflow
- Tests run on push to main and pull requests
